src/status_im2/contexts/chat/composer/link_preview/events.cljs

```diff
@@ -0,0 +1,129 @@
+(ns status-im2.contexts.chat.composer.link-preview.events
+  (:require
+    [clojure.set :as set]
+    [clojure.string :as string]
+    [status-im.data-store.messages :as data-store.messages]
+    [taoensso.timbre :as log]
+    [utils.collection]
+    [utils.re-frame :as rf]))
+
+(rf/defn unfurl-urls
+  {:events [:link-preview/unfurl-urls]}
+  [{:keys [db]} text]
+  (if (string/blank? text)
+    {:db (update db :chat/link-previews dissoc :unfurled :request-id)}
+    {:json-rpc/call
+     [{:method     "wakuext_getTextURLs"
+       :params     [text]
+       :on-success #(rf/dispatch [:link-preview/unfurl-parsed-urls %])
+       :on-error   (fn [error]
+                     (log/error "Failed to parse text and extract URLs"
+                                {:error error
+                                 :event :link-preview/unfurl-urls}))}]}))
+
+(defn- urls->previews
+  [preview-cache urls]
+  (->> urls
+       (map #(get preview-cache % {:url % :loading? true}))
+       (remove :failed?)))
+
+(rf/defn unfurl-parsed-urls
+  {:events [:link-preview/unfurl-parsed-urls]}
+  [{:keys [db]} urls]
+  (let [cleared (set (get-in db [:chat/link-previews :cleared]))]
+    (when (or (not= (set urls) cleared)
+              (empty? urls))
+      (let [cache      (get-in db [:chat/link-previews :cache])
+            previews   (urls->previews cache urls)
+            new-urls   (->> previews
+                            (filter :loading?)
+                            (map :url))
+            ;; `request-id` is a must because we need to process only the last
+            ;; unfurling event, as well as avoid needlessly updating the app db
+            ;; if the user changes the URLs in the input text when there are
+            ;; in-flight RPC requests.
+            request-id (str (random-uuid))]
```

Comment:
Just trying to explain again: this `request-id` technique is necessary when we need only the last success event handler to win. And this is necessary because if the user types a URL `A` and then types another URL `B`, `A` can actually arrive after `B` because both requests to status-go happen asynchronously. I have faced this issue during development a number of times.